### PR TITLE
Reinitiation of lost Child SAs

### PR DESCRIPTION
**Pull request by emilingerslev**
Lots of work went into this PR, but the important part is the `--enable-reinitiator` part.
This feature allows strong-duckling to initiate the missing connection when it's not there.

To make this PR there went work into:
- possible better examples, which are focused on testing cases
- one example for reconnection
- solving issues with our vici implementation
- introducing multiple IKESAStatusReceivers (previously known as reporter(s))

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -36,12 +36,18 @@ func main() {
...
 	whoopingAddress := flags.Flag("whooping", "Address on which to start whooping.").String()
 	tcpCheckerAddresses := flags.Flag("tcp-checker", "TCP address to check. Supports <address>:<port> or <name>:<address>:<port>").Strings()
+	enableReinitiator := flags.Flag("enable-reinitiator", "Enables re-initiation of connections when expected Security Associations are missing").Bool()
 	log.AddFlags(flags)
 	flags.HelpFlag.Short('h')
 	flags.Version(version)
 	socket := flags.Flag("vici-socket", "VICI (charon.vici) socket to connect to. Usually /var/run/charon.vici").String()
 	kingpin.MustParse(flags.Parse(os.Args[1:]))
 
+	if *enableReinitiator && *socket == "" {
```
Comment (Crevil) [inline main.go:46]: Just to be consitent can we use either `len(str) == 0` or `str == ""`. In line 148 we use the former. :) 
```suggestion
	if *enableReinitiator && len(*socket) == 0 {
```

File: strongswan/Dockerfile [hunk 2]
```diff
@@ -12,11 +12,17 @@ RUN \
...
   wget \
   curl \
-  supervisor
+  supervisor \
+  git
 
 RUN curl -sL https://deb.nodesource.com/setup_12.x | bash - &&\
   apt-get install -y nodejs
 RUN npm install nodemon -g
+#RUN npm install gomplate -g
```
Comment (Crevil) [inline strongswan/Dockerfile:21]: Could be dropped?

```suggestion
```

File: internal/strongswan/initiate.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		if childSA.State == nil {
+			initiate := initiateData{
+				IKEName:   ikeSAStatus.Name,
+				ChildName: childSA.Name,
+			}
+			select {
+			case i.initiateWorkerChannel <- initiate:
+				// initiation started
+			default:
+				if loggingTime, ok := i.loggingTime[initiate.getFullName()]; !ok || time.Now().Sub(loggingTime) >= 30*time.Second {
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/strongswan/initiate.go:43]: Thats the idea yes 👍  Could be a simple method yes, but it should it also save the new timestamp inside that method then?

File: internal/strongswan/initiate.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		client:                client,
+		logger:                logger,
+		initiateWorkerChannel: initiateWorkerChannel,
+		loggingTime:           map[string]time.Time{},
+	}
+}
+
+func (i *Reinitiator) IKESAStatus(ikeSAStatus IKESAStatus) {
+	for _, childSA := range ikeSAStatus.ChildSA {
+		if childSA.State == nil {
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/strongswan/initiate.go:34]: most def 👍

File: internal/vici/clientConn.go [hunk 5]
```diff
@@ -53,7 +53,7 @@ func NewClientConnFromDefaultSocket() (*ClientConn, error) {
 func (c *ClientConn) Request(apiname string, concretePayload interface{}) (map[string]interface{}, error) {
 	var request map[string]interface{}
 	if concretePayload != nil {
-		err := convertToGeneral(concretePayload, request)
+		err := convertToGeneral(concretePayload, &request)
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/vici/clientConn.go:56]: no, will check 👍

File: strongswan/Dockerfile (hunk 2 above)
Comment (emilingerslev) [reply strongswan/Dockerfile:21]: of cause 👍

File: main.go (hunk 1 above)
Comment (emilingerslev) [reply main.go:46]: 👍

File: internal/vici/initiate.go [hunk 6]
```diff
@@ -6,20 +6,37 @@ import (
...
 		request["ike"] = ike
 	}
+
+	err = c.RegisterEvent("control-log", func(response map[string]interface{}) {
+		logger(response)
+	})
+	if err != nil {
+		return fmt.Errorf("error registering control-log event: %w", err)
+	}
+
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/vici/initiate.go:24]: fixed 🦖

File: internal/vici/clientConn.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/vici/clientConn.go:56]: only place its used 😮 🤔 👍

[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/vici/clientConn.go:56]: Ah, its in the shared `Request` function. 👍

File: internal/strongswan/initiate.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+var _ IKESAStatusReceiver = &Reinitiator{}
+
+type Reinitiator struct {
+	client                *vici.ClientConn
+	logger                log.Logger
+	initiateWorkerChannel chan initiateData
+	loggingTime           map[string]time.Time
+	currentInitiate       initiateData
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/strongswan/initiate.go:18]: Always just set it in the same method and don't unset it, so we should be fine without a mutex. However I forgot to set it, so will fix that.

File: internal/strongswan/initiate.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/strongswan/initiate.go]: Decided to do the function, but keeping the `i.loggingTime[initiate.getFullName()] = time.Now()`

File: internal/vici/clientConn.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Crevil) [reply internal/vici/clientConn.go:56]: 👍

**Pull request merged**